python-template/{{cookiecutter.placeholder_repo_name}}/.github/PULL_REQUEST_TEMPLATE.md

```diff
@@ -1,39 +1,8 @@
-**Description:** Describe in a couple of sentences what this PR adds
-
-**JIRA:** Link to JIRA ticket
-
-**Dependencies:** dependencies on other outstanding PRs, issues, etc. 
-
-**Merge deadline:** List merge deadline (if any)
-
-**Installation instructions:** List any non-trivial installation 
-instructions.
-
-**Testing instructions:**
-
-1. Open page A
-2. Do thing B
-3. Expect C to happen
-4. If D happened instead - check failed.
-
-**Reviewers:**
-- [ ] tag reviewer 
-- [ ] tag reviewer 
 
 **Merge checklist:**
-- [ ] All reviewers approved
-- [ ] CI build is green
+Check off if complete or not applicable:
```

Comment:
Sorry, I thought of one more thing 😬 

```suggestion
Check off if complete or not applicable:
- [ ] Backported to [latest and next named releases](https://docs.openedx.org/en/latest/community/release_notes/named_release_branches_and_tags.html?highlight=latest#latest-open-edx-release)
```

IMO that would remove the need for BTR to stick [this big warning at the top of the edx-platform PR template](https://github.com/openedx/edx-platform/blob/674bf2a040aa97fd2b98b8a513b0ec592fbeec8f/.github/pull_request_template.md?plain=1#L1-L20).

Reply:
Backports are usually only relevant for IDAs, right? And versioning and changelogs are usually only done for libraries. So maybe it makes sense to keep separate templates for the various cookiecutters after all.

Reply:
It's a PITA so we try to avoid it, but we do need to backport fixes to libraries sometimes: https://openedx.atlassian.net/wiki/spaces/COMM/pages/2065367719/Making+a+pull+request+for+a+named+release#Backporting-to-a-library-repository

But anyway, to completely contradict my own suggestion: maybe this is trying to hard? I'm realizing that any truly exhaustive list of pre-merge items would be too long for a good PR template. I think there's virtue in having one simple, universal PR template, rather than trying to build a bunch of templates for different types of repos.

In terms of versioning and changelogs: Can we phrase it in a way that works for IDAs too? Maybe something like:

> - [Release process]() followed

Where "release process" is a link to a wiki page that explains or links out to:
* our backend library release process (version bump, changelog, whatever)
* our frontend library release process (version bump, GH release, whatever)
* the process for MFEs and IDAs, for:
  * core contributors (message relevant `#cc-*` channel)
  * 2U engineers (could link to an internal page)
* the backporting process & how to know whether it applies to your PR

I realize that building such a page would take some work...

Reply:
After more consideration, I think we actually want to go in the exact opposite direction: Customize each repo's PR template as much as possible. Some repos keep changelogs, others do not; some have versions, others are continuously released; for some, manual testing is appropriate, and for others it is not.... The more I tried to create a universal template, the larger it got.

Fundamentally, I think there are three critical principles in checklist design:

1. Make it as short as possible -- otherwise people get annoyed/fatigued/intimidated
2. Have zero irrelevant items for the vast majority of occasions -- otherwise people get in the habit of checking off items without considering them carefully
3. Have the easiest-to-forget things highest in the list, to increase the chance they are noticed

I don't think you can create a universal template and also stick to those items, particularly # 2. In fact, not having the template be too uniform probably hurts # 3, as it increases a sense of familiarity, which reduces the chance people will stop and think.

Anyway, take a look at the latest. The commit message describes the differences between the variations. If you hate it, I can back it out, but I think this approach has the best chance of increasing participation.

Reply:
I agree with those checklist design principles. Once the checklists are designed, though, I think three things need to happen in order for them to be adopted:

1. **New repos get the correct PR template.** This is easy enough.
2. **Existing repositories are updated to use correct PR template.** (Yes, we could skip on updating existing repos. However, I don't see these checklists having an impact unless folks see them basically everywhere.)
   * For an org-wide template, it'd be easy. Just delete all existing PR templates 💥 
   * For customized templates, you'd need to go in and update every repo to the right PR template.
3. **Checklists are updated as repos & processes change over time.**
   * For an org-wide template, this would just mean editing the central PR template or supplementary wiki pages.
   * For customized templates, this would mean updating templates across repositories, every time things change. Either maintainers or a central team like Arch-BOM would need to take on this responsibility. (FWIW, tCRIL has [some new automation for cross-repo updates](https://github.com/openedx/terraform-github/blob/main/migrate/repo_checks.py). We've been using it to keep certain workflows consistent across repos. Still, merging all the generated PRs is annoying).

Personally, I think I would still choose an org-wide template just to avoid taking on the up-front cost of (2) and the maintenance burden of (3). That said, if customized templates seem worth the lift to you, then no opposition from me 👍🏻

Reply:
Yeah, I definitely feel the tension of "ugh, not another thing to keep updated", which I think I actually mentioned earlier. But PR templates are intimately tied to ownership, and different maintainers have different checklists (at least in their minds, if not explicitly written down.) It would actually be really cool if we could standardize maintainership, and then I think there'd be a really strong argument for standardized PR template checklists. But even then we'd have differences for libraries vs. IDAs, and GH doesn't have a way to say "I want Standard PR Template No. 5".

I'm OK with the idea of having to automate (or semi-automate) replacement of PR templates at some point. I think we're going to need to get better at that sort of thing eventually anyhow. One option would be to *fully* automate modification of PR templates that haven't been edited; that would be a pretty safe thing to do, since you could say "auto-merge any where the the diff looks *exactly* like this" and be confident that the automation wasn't screwing anything up. That's how I'd want to approach any such future change.

Reply:
Sounds good 👍

Once you merge I recommend linking up with the #maintainership-pilot folks. They have a draft repo standards list going and it sounds like "appropriate PR template" ought to be on there.